### PR TITLE
HBASE-26759 Fix trace continuity through CallRunner

### PR DESCRIPTION
**Pull request by ndimiduk**
We appear to be generating spans with without a parent context with the name
`RegionScanner.close`.

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m  5s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   2m 56s |  master passed  |
| +1 :green_heart: |  compile  |   1m 58s |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 36s |  master passed  |
| +1 :green_heart: |  spotbugs  |   1m 22s |  master passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   2m 28s |  the patch passed  |
| +1 :green_heart: |  compile  |   2m  2s |  the patch passed  |
| +1 :green_heart: |  javac  |   2m  2s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m 38s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  hadoopcheck  |  12m 27s |  Patch does not cause any errors with Hadoop 3.1.2 3.2.2 3.3.1.  |
| +1 :green_heart: |  spotbugs  |   1m 29s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 11s |  The patch does not generate ASF License warnings.  |
|  |   |  32m 47s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4126/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/4126 |
| Optional Tests | dupname asflicense javac spotbugs hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 7c897c512850 5.4.0-1025-aws #25~18.04.1-Ubuntu SMP Fri Sep 11 12:03:04 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / cd45cadbc1 |
| Default Java | AdoptOpenJDK-1.8.0_282-b08 |
| Max. process+thread count | 60 (vs. ulimit of 30000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4126/1/console |
| versions | git=2.17.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by ndimiduk** [comment]
I did not attempt to perform an audit of the span lifecycle, only reconnected the orphaned span to what appears to be its proper parent. If you think there are bugs in properly closing spans, I hope that we can find them by way of the pom change i have over on https://github.com/apache/hbase/pull/4106/files#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8

**Comment by ndimiduk** [comment]
I do find it strange that we have a span explicitly for a `close` method, but not other lifecycle methods on the same object (such as `open`, for example).

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 37s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  2s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   2m 45s |  master passed  |
| +1 :green_heart: |  compile  |   0m 34s |  master passed  |
| +1 :green_heart: |  shadedjars  |   5m  8s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 24s |  master passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   2m 24s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 35s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 35s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   5m 12s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 22s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 327m 43s |  hbase-server in the patch passed.  |
|  |   | 347m 22s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4126/1/artifact/yetus-jdk8-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/4126 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux edaf4ac37bb9 5.4.0-1025-aws #25~18.04.1-Ubuntu SMP Fri Sep 11 12:03:04 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / cd45cadbc1 |
| Default Java | AdoptOpenJDK-1.8.0_282-b08 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4126/1/testReport/ |
| Max. process+thread count | 1727 (vs. ulimit of 30000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4126/1/console |
| versions | git=2.17.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache9** [comment]
I mean we have a span here

https://github.com/apache/hbase/blob/5dc663ea388d64f9d583a9e8151da6b8d403b0fd/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/ServerRpcConnection.java#L634

I think it will be the parent span for the spans at server side, and seems it has not been ended properly?

**Comment by ndimiduk** [comment]
@Apache9 

> I think the problem here is that we forget to end the span created in ServerRpcConnection?

I see that the trace scope is closed properly, but I think you're correct that the trace is not ended, neither with success nor failure.

> Anyway, I think we should have a big span to trace the whole rpc processing...

My impression was that this span in `processRequest` is the big span to trace the whole rpc processing. Is there additional scope you would see included?

**Comment by ndimiduk** [comment]
@Apache9 please take another look.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   3m 38s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 48s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m 35s |  master passed  |
| +1 :green_heart: |  compile  |   2m 52s |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 48s |  master passed  |
| +1 :green_heart: |  spotbugs  |   1m 50s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 10s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m 23s |  the patch passed  |
| +1 :green_heart: |  compile  |   2m 47s |  the patch passed  |
| +1 :green_heart: |  javac  |   2m 47s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m 13s |  The patch passed checkstyle in hbase-client  |
| +1 :green_heart: |  checkstyle  |   0m 33s |  hbase-server: The patch generated 0 new + 20 unchanged - 2 fixed = 20 total (was 22)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  hadoopcheck  |  11m 50s |  Patch does not cause any errors with Hadoop 3.1.2 3.2.2 3.3.1.  |
| +1 :green_heart: |  spotbugs  |   2m  1s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 16s |  The patch does not generate ASF License warnings.  |
|  |   |  39m  6s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4126/2/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/4126 |
| Optional Tests | dupname asflicense javac spotbugs hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 9b8727bc845d 5.4.0-1043-aws #45~18.04.1-Ubuntu SMP Fri Apr 9 23:32:25 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / 9c037b1be7 |
| Default Java | AdoptOpenJDK-1.8.0_282-b08 |
| Max. process+thread count | 60 (vs. ulimit of 30000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4126/2/console |
| versions | git=2.17.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache9** [comment]
> @Apache9
> 
> > I think the problem here is that we forget to end the span created in ServerRpcConnection?
> 
> I see that the trace scope is closed properly, but I think you're correct that the trace is not ended, neither with success nor failure.
> 
> > Anyway, I think we should have a big span to trace the whole rpc processing...
> 
> My impression was that this span in `processRequest` is the big span to trace the whole rpc processing. Is there additional scope you would see included?

Ah, sorry, forgot to reply here. No additional scope. Let me check the newest PR.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/ServerRpcConnection.java [hunk 1]
```diff
@@ -716,14 +717,22 @@ public String get(RPCTInfo carrier, String key) {
       ServerCall<?> call = createCall(id, this.service, md, header, param, cellScanner,
         totalRequestSize, this.addr, timeout, this.callCleanup);
 
-      if (!this.rpcServer.scheduler.dispatch(new CallRunner(this.rpcServer, call))) {
+      if (this.rpcServer.scheduler.dispatch(new CallRunner(this.rpcServer, call, span))) {
+        // unset span do that it's not closed in the finally block
+        span = null;
```
Comment (Apache9) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/ServerRpcConnection.java:722]: Yes, the code structure here is not friendly for tracing...

Thanks for the detailed comments!

**Review by Apache9 — APPROVED** [review]
+1

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 36s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m  9s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   2m 21s |  master passed  |
| +1 :green_heart: |  compile  |   2m 48s |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 47s |  master passed  |
| +1 :green_heart: |  spotbugs  |   1m 47s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m  9s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m 19s |  the patch passed  |
| +1 :green_heart: |  compile  |   2m 53s |  the patch passed  |
| +1 :green_heart: |  javac  |   2m 53s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m 14s |  The patch passed checkstyle in hbase-client  |
| +1 :green_heart: |  checkstyle  |   0m 34s |  hbase-server: The patch generated 0 new + 20 unchanged - 2 fixed = 20 total (was 22)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  hadoopcheck  |  11m 29s |  Patch does not cause any errors with Hadoop 3.1.2 3.2.2 3.3.1.  |
| +1 :green_heart: |  spotbugs  |   2m  2s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 14s |  The patch does not generate ASF License warnings.  |
|  |   |  33m 42s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4126/3/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/4126 |
| Optional Tests | dupname asflicense javac spotbugs hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 708f764fbc78 5.4.0-1043-aws #45~18.04.1-Ubuntu SMP Fri Apr 9 23:32:25 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / f4b05a8364 |
| Default Java | AdoptOpenJDK-1.8.0_282-b08 |
| Max. process+thread count | 60 (vs. ulimit of 30000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4126/3/console |
| versions | git=2.17.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 41s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  2s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 13s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   2m 41s |  master passed  |
| +1 :green_heart: |  compile  |   0m 49s |  master passed  |
| +1 :green_heart: |  shadedjars  |   4m  0s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 35s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 11s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m 18s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 50s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 50s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   4m  1s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 34s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 36s |  hbase-client in the patch passed.  |
| -1 :x: |  unit  | 324m  8s |  hbase-server in the patch failed.  |
|  |   | 344m 33s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4126/3/artifact/yetus-jdk8-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/4126 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux e46f4ae75c60 5.4.0-1025-aws #25~18.04.1-Ubuntu SMP Fri Sep 11 12:03:04 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / f4b05a8364 |
| Default Java | AdoptOpenJDK-1.8.0_282-b08 |
| unit | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4126/3/artifact/yetus-jdk8-hadoop3-check/output/patch-unit-hbase-server.txt |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4126/3/testReport/ |
| Max. process+thread count | 1796 (vs. ulimit of 30000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4126/3/console |
| versions | git=2.17.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 38s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  2s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 36s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m 29s |  master passed  |
| +1 :green_heart: |  compile  |   1m  3s |  master passed  |
| +1 :green_heart: |  shadedjars  |   4m 17s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 41s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 11s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m  2s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  4s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m  4s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   4m 11s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 40s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 52s |  hbase-client in the patch passed.  |
| +1 :green_heart: |  unit  | 333m 19s |  hbase-server in the patch passed.  |
|  |   | 357m 37s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4126/3/artifact/yetus-jdk11-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/4126 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux ba4d8893f58b 5.4.0-1025-aws #25~18.04.1-Ubuntu SMP Fri Sep 11 12:03:04 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / f4b05a8364 |
| Default Java | AdoptOpenJDK-11.0.10+9 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4126/3/testReport/ |
| Max. process+thread count | 1577 (vs. ulimit of 30000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4126/3/console |
| versions | git=2.17.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Pull request merged**